File: sandbox/target-perf-dotnet/Inefficient.cs
// Deliberately inefficient C# code for rule-validation only.
// Each method maps 1:1 to a rule in orchestrator/rules/efficiency/dotnet-*.yml.
// DO NOT use in production. Wired for static analysis only -- no Program.cs,
// no .csproj; Semgrep parses these files standalone.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace TargetPerfDotnet;

public static class Inefficient
{
    // efficiency.dotnet.linear-search-in-loop
    public static List<int> FindCommonAccountIds(List<int> a, List<int> b)
    {
        var common = new List<int>();
        foreach (var id in a)
        {
            if (b.Contains(id))   // O(n) per probe -> O(n*m)
            {
                common.Add(id);
            }
        }
        return common;
    }

    // efficiency.dotnet.nested-loop-same-collection
    public static List<(string, string)> FindDuplicatePairs(List<Transaction> txs)
    {
        var pairs = new List<(string, string)>();
        foreach (var a in txs)
        {
            foreach (var b in txs)     // O(n^2)
            {
                if (a.Id != b.Id && a.UserId == b.UserId && a.Amount == b.Amount)
                {
                    pairs.Add((a.Id, b.Id));
                }
            }
        }
        return pairs;
    }

    // efficiency.dotnet.string-concat-in-loop
    public static string BuildCsvExport(List<string[]> rows)
    {
        string output = "";
        foreach (var row in rows)
        {
            output += string.Join(",", row) + "\n";   // O(n^2) memory churn
        }
        return output;
    }

    // efficiency.dotnet.regex-recompile-in-loop
    public static List<string> MaskPansInLogLines(List<string> lines)
    {
        var masked = new List<string>();
        foreach (var line in lines)
        {
            var pattern = new Regex(@"\b(\d{6})\d{6}(\d{4})\b");  // re-compiled per line
            masked.Add(pattern.Replace(line, "$1******$2"));
        }
        return masked;
    }

    // efficiency.dotnet.linq-count-vs-any
    public static bool HasAnyApprovedPayment(IEnumerable<Transaction> txs)
    {
        return txs.Where(t => t.Status == "approved").Count() > 0;  // .Any() is cheaper
    }

    // efficiency.dotnet.tolist-before-filter
    public static List<Transaction> RecentLargeTransactions(IQueryable<Transaction> source)
    {
        return source.ToList().Where(t => t.Amount > 1000m).ToList();
        // materializes the whole table before filtering
    }
}

public class Transaction
{
    public string Id { get; set; } = "";
    public int UserId { get; set; }
    public decimal Amount { get; set; }
    public string Status { get; set; } = "";
}

// Stub so the file is self-contained from Semgrep's POV; not actually used.
public interface IQueryable<T> : IEnumerable<T> { }

File: sandbox/target-perf-java/Inefficient.java
// Deliberately inefficient Java code for rule-validation only.
// Each method maps 1:1 to a rule in orchestrator/rules/efficiency/java-*.yml.
// DO NOT use in production. Wired for static analysis only -- no main(),
// no build.gradle; Semgrep parses this file standalone.

package com.aci.speedpay.perf;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class Inefficient {

    // efficiency.java.linear-search-in-loop
    public static List<Integer> findCommonAccountIds(List<Integer> a, List<Integer> b) {
        List<Integer> common = new ArrayList<>();
        for (Integer id : a) {
            if (b.contains(id)) {          // O(n) per probe -> O(n*m)
                common.add(id);
            }
        }
        return common;
    }

    // efficiency.java.nested-loop-same-collection
    public static List<String> findDuplicatePairs(List<Transaction> txs) {
        List<String> pairs = new ArrayList<>();
        for (Transaction a : txs) {
            for (Transaction b : txs) {                    // O(n^2)
                if (!a.id.equals(b.id) && a.userId == b.userId && a.amount == b.amount) {
                    pairs.add(a.id + "+" + b.id);
                }
            }
        }
        return pairs;
    }

    // efficiency.java.string-concat-in-loop
    public static String buildCsvExport(List<String[]> rows) {
        String output = "";
        for (String[] row : rows) {
            output += String.join(",", row) + "\n";        // O(n^2) memory churn
        }
        return output;
    }

    // efficiency.java.regex-recompile-in-loop
    public static List<String> maskPansInLogLines(List<String> lines) {
        List<String> masked = new ArrayList<>();
        for (String line : lines) {
            Pattern pattern = Pattern.compile("\\b(\\d{6})\\d{6}(\\d{4})\\b");
            masked.add(pattern.matcher(line).replaceAll("$1******$2"));
        }
        return masked;
    }

    // efficiency.java.boxing-in-arithmetic-loop
    public static Integer totalAmountCents(List<Transaction> txs) {
        Integer total = 0;                                  // boxed accumulator
        for (Transaction t : txs) {
            total = total + (int)(t.amount * 100);          // autobox each iteration
        }
        return total;
    }

    // Domain stub
    public static class Transaction {
        public String id = "";
        public int userId;
        public double amount;
        public String status = "";
    }
}
